Было:
`=== RUN   TestGetDomainStat_Time_And_Memory
    stats_optimization_test.go:46: time used: 383.073021ms / 300ms
    stats_optimization_test.go:47: memory used: 308Mb / 30Mb
    assertion_compare.go:332: 
                Error Trace:    stats_optimization_test.go:49
                Error:          "383073021" is not less than "300000000"
                Test:           TestGetDomainStat_Time_And_Memory
                Messages:       [the program is too slow]
--- FAIL: TestGetDomainStat_Time_And_Memory (6.21s)
FAIL
FAIL    github.com/uVazzi/go-otus/hw10_program_optimization     6.212s
FAIL`

Стало:
`=== RUN   TestGetDomainStat_Time_And_Memory
    stats_optimization_test.go:46: time used: 143.828233ms / 300ms
    stats_optimization_test.go:47: memory used: 11Mb / 30Mb
--- PASS: TestGetDomainStat_Time_And_Memory (1.58s)
PASS
ok      github.com/uVazzi/go-otus/hw10_program_optimization     1.582s`

--- a/hw10_program_optimization/stats.go
+++ b/hw10_program_optimization/stats.go
@@ -3,7 +3,6 @@ package hw10programoptimization
 import (
 	"encoding/json"
 	"errors"
-	"fmt"
 	"io"
 	"strings"
 )
@@ -17,34 +16,17 @@ type DomainStat map[string]int
 var ErrIncorrectEmail = errors.New("incorrect email")
 
 func GetDomainStat(r io.Reader, domain string) (DomainStat, error) {
-	u, err := getUsers(r)
-	if err != nil {
-		return nil, fmt.Errorf("get users error: %w", err)
-	}
-	return countDomains(u, domain)
-}
-
-type users []User
-
-func getUsers(r io.Reader) (result users, err error) {
 	var user User
+	result := make(DomainStat)
 	decoder := json.NewDecoder(r)
 
 	for decoder.More() {
-		err = decoder.Decode(&user)
+		err := decoder.Decode(&user)
 		if err != nil {
-			return
+			return nil, err
 		}
-		result = append(result, user)
-	}
 
-	return
-}
-
-func countDomains(usersData users, domain string) (DomainStat, error) {
-	result := make(DomainStat)
-	for _, user := range usersData {
-		if strings.HasSuffix(user.Email, domain) {
+		if strings.HasSuffix(user.Email, "."+domain) {
 			key := strings.SplitN(user.Email, "@", 2)
 			if len(key) != 2 {
 				return nil, ErrIncorrectEmail

